snuba/migrations/snuba_migrations/events/0016_drop_legacy_events.py

```diff
@@ -0,0 +1,191 @@
+    Column("http_method", String(Modifiers(nullable=True))),
+    Column("http_referer", String(Modifiers(nullable=True))),
+    Column(
+        "exception_stacks",
+        Nested(
+            [
+                ("type", String(Modifiers(nullable=True))),
+                ("value", String(Modifiers(nullable=True))),
+                ("mechanism_type", String(Modifiers(nullable=True))),
+                ("mechanism_handled", UInt(8, Modifiers(nullable=True))),
+            ]
+        ),
+    ),
+    Column(
+        "exception_frames",
+        Nested(
+            [
+                ("abs_path", String(Modifiers(nullable=True))),
+                ("filename", String(Modifiers(nullable=True))),
+                ("package", String(Modifiers(nullable=True))),
+                ("module", String(Modifiers(nullable=True))),
+                ("function", String(Modifiers(nullable=True))),
+                ("in_app", UInt(8, Modifiers(nullable=True))),
+                ("colno", UInt(32, Modifiers(nullable=True))),
+                ("lineno", UInt(32, Modifiers(nullable=True))),
+                ("stack_level", UInt(16)),
+            ]
+        ),
+    ),
+    Column("culprit", String(Modifiers(nullable=True))),
+    Column("sdk_integrations", Array(String())),
+    Column("modules", Nested([("name", String()), ("version", String())])),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    """
+    This is exactly the same as 0001_events_initial with the forwards and backwards
+    methods flipped around since we are deleting the table now.
+    """
+
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.EVENTS, table_name="sentry_local"
+            )
+        ]
+
+    def backwards_local(self) -> Sequence[operations.SqlOperation]:
+        sample_expr = "cityHash64(toString(event_id))"
+
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.EVENTS,
+                table_name="sentry_local",
+                columns=columns,
+                engine=table_engines.ReplacingMergeTree(
+                    storage_set=StorageSetKey.EVENTS,
+                    version_column="deleted",
+                    order_by="(project_id, toStartOfDay(timestamp), %s)" % sample_expr,
+                    partition_by="(toMonday(timestamp), if(equals(retention_days, 30), 30, 90))",
+                    sample_by=sample_expr,
+                ),
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.EVENTS,
+                table_name="sentry_local",
+                column=Column(
+                    "_tags_hash_map",
+                    Array(UInt(64), Modifiers(materialized=TAGS_HASH_MAP_COLUMN)),
+                ),
+                after="_tags_flattened",
+            ),
+        ]
```

Comment:
This seems like a lot of effort to support backwards_local. Having to define all the columns again for supporting reverts. And there could be instances where you might miss a column which might have been added by a migration. Is there a mechanism in migrations where you can provide Clickhouse SQL directly? That would be useful in such scenarios

Reply:
Yes there is. It's `operations.RunSql`. I considered just dumping the strings there but I did it this way because I (originally) thought it was a simple copy and paste.

Reply:
My primary concern is that we could have added/removed columns, added bloom filter indexes, modified TTL's after the first migration. A copy/paste of just the columns would miss them. Can you verify that the SQL command created by this is similar to the show create command on clickhouse?

Reply:
Yes, it is the exact same string when reversing as it is before the migration. I've checked all the migrations for the things you described.